### ISSUE TITLE
Plugin concurrency issue

### ISSUE THREAD

**Issue by tessereth** [buildkite/agent#763]
I’ve run into a concurrency bug with plugins. I have two steps, run concurrently, both using the same plugin. We also run two agents per box. If the two steps are picked up by agents on the same box, the first sees the plugin doesn’t exist and clones it. The second sees the directory exists, tries to continue but fails to find the command script in the plugin. You probably need some kind of locking or something though I haven’t looked into it too hard. If you need any additional information, just let me know.

Agent version: v3.1.1

First step:

```
# Plugin "github.com/bugcrowd/test-summary-buildkite-plugin" will be checked out to "/etc/buildkite-agent/plugins/github-com-bugcrowd-test-summary-buildkite-plugin-9637cfd5e8383996c09f6e462c4f664d1a9e612a"
$ cd /etc/buildkite-agent/plugins/github-com-bugcrowd-test-summary-buildkite-plugin-9637cfd5e8383996c09f6e462c4f664d1a9e612a
# Switching to the plugin directory
$ git clone -v -- https://github.com/bugcrowd/test-summary-buildkite-plugin .
Cloning into '.'...
POST git-upload-pack (590 bytes)
remote: Counting objects: 650, done.
remote: Compressing objects: 100% (44/44), done.
remote: Total 650 (delta 16), reused 27 (delta 7), pack-reused 597
Receiving objects: 100% (650/650), 1.30 MiB | 0 bytes/s, done.
Resolving deltas: 100% (362/362), done.
Checking connectivity... done.
# Checking out `9637cfd5e8383996c09f6e462c4f664d1a9e612a`
$ git checkout -f 9637cfd5e8383996c09f6e462c4f664d1a9e612a
Note: checking out '9637cfd5e8383996c09f6e462c4f664d1a9e612a'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

 git checkout -b <new-branch-name>

HEAD is now at 9637cfd... Return zero exit status on error
$ cd /

Preparing working directory	0s
$ cd /var/lib/buildkite-agent/builds/[REDACTED]/test-summary-buildkite-plugin
# Host "github.com" already in list of known hosts at "/var/lib/buildkite-agent/.ssh/known_hosts"
$ git remote set-url origin git@github.com:bugcrowd/test-summary-buildkite-plugin.git
$ git clean -fxdq
$ git fetch -v origin 9637cfd5e8383996c09f6e462c4f664d1a9e612a
From github.com:bugcrowd/test-summary-buildkite-plugin
* branch            9637cfd5e8383996c09f6e462c4f664d1a9e612a -> FETCH_HEAD
$ git checkout -f 9637cfd5e8383996c09f6e462c4f664d1a9e612a
HEAD is now at 9637cfd... Return zero exit status on error
# Cleaning again to catch any post-checkout changes
$ git clean -fxdq
# Checking to see if Git data needs to be sent to Buildkite
$ buildkite-agent meta-data exists buildkite:git:commit
```

Second step:

```
Setting up plugins	0s
# Plugin "github.com/bugcrowd/test-summary-buildkite-plugin#9637cfd5e8383996c09f6e462c4f664d1a9e612a" already checked out (can't `git rev-parse HEAD` plugin git directory)
Preparing working directory	0s
$ cd /var/lib/buildkite-agent/builds/[REDACTED]/test-summary-buildkite-plugin
# Host "github.com" already in list of known hosts at "/var/lib/buildkite-agent/.ssh/known_hosts"
$ git remote set-url origin git@github.com:bugcrowd/test-summary-buildkite-plugin.git
$ git clean -fxdq
$ git fetch -v origin 9637cfd5e8383996c09f6e462c4f664d1a9e612a
From github.com:bugcrowd/test-summary-buildkite-plugin
* branch            9637cfd5e8383996c09f6e462c4f664d1a9e612a -> FETCH_HEAD
$ git checkout -f 9637cfd5e8383996c09f6e462c4f664d1a9e612a
HEAD is now at 9637cfd... Return zero exit status on error
# Cleaning again to catch any post-checkout changes
$ git clean -fxdq
# Checking to see if Git data needs to be sent to Buildkite
$ buildkite-agent meta-data exists buildkite:git:commit
🚨 Error: No command has been provided
```

**Comment by sj26** [comment]
Oh no, it looks like there's a race condition in plugin checkout. Once the directory exists, the agent just assumes it's checked out and ready to go. I'm guessing your steps ran at the same time on the same host with the same plugin directory, and one was checking out the plugin while the other one plowed ahead, and so the plugin hooks weren't available. It looks like we take a lock, but too late.

https://github.com/buildkite/agent/blob/master/bootstrap/bootstrap.go#L480-L507

@lox can we take the lock earlier? I think we'll need a top-level lock around checking for the directory and making the directory, too.

**Comment by lox** [comment]
Will see what I can do!

**Comment by lox** [comment]
Fixed in 3.5.0!

**Comment by tessereth** [comment]
Sweet thanks!
